CDCgov/prime-reportstream
# 8642: return more informative error messages on authenticate errors

This PR introduces a mechanism and proposal for surfacing authentication errors to end users

Test Steps:
There are a bunch of automated tests that cover this, but to verify this locally there are several of ways to generate an error using the generate-jwt.py script:
```sh
curl -X POST -H "Content-Type: application/x-www-form-urlencoded"  -d "scope=healthy-labs.*.report&grant_type=client_credentials&client_assertion_type=urn:ietf:params:oauth:client-assertion-type:jwt-bearer&client_assertion=eyJhbGciOiJSUzI1NiIsImtpZCI6ImRlZmF1bHQiLCJ0eXAiOiJKV1QifQ.eyJpc3MiOiJoZWFsdGh5LWxhYnMiLCJzdWIiOiJoZWFsdGh5LWxhYnMiLCJhdWQiOiJzdGFnaW5nLnByaW1lLmNkYy5nb3YiLCJleHAiOjE2ODM4MTE2NjIsImp0aSI6IjVmOTBlYTkxLTc0NGEtNDM5ZC1iNjEzLTAyNTAyNTVjMDg2YSJ9.UNzNHtCidxTTpE9XISxeSCJmeqFpr5VVJV145OrqCcW-_w9FtWGAot1fIHt84A5fNuOS_2ce4F2jgk8htEO-tuX_Uik522GC_TnBaawtUslKhFMSJXS1g1daqpgaTSRle84nXmSm9Cgho2GLrk385GQM3-m9gHkZA0I7ZBEAptRWCgJVfbUE6jVxqeEFtuoRd6frB5gcY-zh2mUTl0ONgoF0wMxPqGYM2osTOVZr3wfEWlyITx6FMzGURXkHjS1OZz05kqORvd-SUNxO6f13ikTNCWIay_I5qt7DFC1o1WeMxOGvt4zI7c8qaicNPqVSDUCvH3NMjHW0R-dWI_aMmQ" "http://localhost:7071/api/token" 
```

- create a JWT, wait ten minutes and request an access token with it
- create a jwt, but set the scope in the script to something bogus like `foo.bar`
- create a jwt, but set the `my_client_id` param to an org that does not exist
- create a jwt, but set the `kid` param to one that does not match any of the keys
- create a jwt, request a token and then request a token again with the same jwt

## Changes
- Updates the auth code to return the correct error for auth issue encountered and an `error_uri`
- Adds a proposal for how the `error_uri` would translate to a page on the site

## Checklist

### Testing

### Process

## Linked Issues
- Fixes #issue

## To Be Done
- There will be follow up work with the webex team to add the new web resources

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
